src/core/net/ip6.cpp

```diff
@@ -1492,6 +1502,66 @@ bool Ip6::IsOnLink(const Address &aAddress) const
     return rval;
 }
 
+void Ip6::UpdateBorderRoutingCounters(const Message &  aMessage,
+                                      const Header &   aHeader,
+                                      const Header *   aInnerHeader,
+                                      bool             aForwardHost,
+                                      bool             aForwardThread,
+                                      const OptionMpl &aMplOption)
+{
+    otPacketsAndBytes *counter = nullptr;
+
+    if (!Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetSource()) &&
+        !Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetDestination()) &&
+        !aHeader.GetSource().IsLinkLocal() && !aHeader.GetDestination().IsLinkLocal() &&
+        aHeader.GetSource().GetPrefix() != Get<Mle::Mle>().GetMeshLocalPrefix() &&
+        aHeader.GetDestination().GetPrefix() != Get<Mle::Mle>().GetMeshLocalPrefix())
+    {
+        if (aForwardThread)
+        {
+            if (aHeader.GetDestination().IsMulticast())
+            {
+                // Do nothing. Inbound Multicast is handled later.
+            }
+            else
+            {
+                counter = &mBorderRoutingCounters.mInboundUnicast;
+            }
+        }
+        if (aForwardHost)
```

Comment:
I would suggest maybe rethinking the model about where/how we check and update the BR counters. It may be best to not use `HandleDatagram()` directly specially since it is called from many places (e.g. from `Ip6::HandleSendQueue()`)

How about the following:
- Everything host sends to us comes through `Ip6::SendRaw()` (which maps to `otIp6Send`), we can use this as place to update inbound counters.
- Everything we pass back to host will go through `Ip6::ProcessReceiveCallback()`, we can use this as place for outbound counter.

Thoughts?